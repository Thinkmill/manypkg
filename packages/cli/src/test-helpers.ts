// These are utils for testing.
// I wanted to put this in the test folder but then jest tested it
// Easier to put it here than add to jest ignore but maybe wrong.
//
// Who can say? ¯\_(ツ)_/¯

import { Package } from "@manypkg/get-packages";
import crypto from "crypto";

export let getFakeWS = (
  name: string = "pkg-1",
  version: string = "1.0.0"
): Package => {
  return {
    dir: `some/fake/dir/${name}`,
    packageJson: {
      name,
      version
    }
  };
};

export let getWS = (): Map<string, Package> => {
  let pkg = new Map();
  pkg.set("pkg-1", getFakeWS());
  return pkg;
};

export let getFakeString = (length: number): string => {
  return (
    crypto
      // converting to hex doubles the length, so we request half as many bytes
      .randomBytes(Math.ceil(length / 2))
      .toString("hex")
      // if length is odd, Math.ceil() will have requested too many bytes, so we
      // chop them off by only grabbing `length` characters
      .substring(-length)
  );
};
