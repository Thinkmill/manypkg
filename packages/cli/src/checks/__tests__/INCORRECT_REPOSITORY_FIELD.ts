import path from "path";
import check from "../INCORRECT_REPOSITORY_FIELD";
import { getWS, getFakeWS, getFakeString } from "../../test-helpers";

describe("incorrect repository field", () => {
  describe("github", () => {
    it("should work", () => {
      let ws = getWS();
      let rootWorkspace = getFakeWS("root");
      let defaultBranch = `b${getFakeString(5)}`;

      (rootWorkspace.packageJson as any).repository =
        "https://github.com/Thinkmill/manypkg";
      rootWorkspace.dir = __dirname;
      let workspace = getFakeWS("no-repository-field");
      workspace.dir = path.join(__dirname, "packages/no-repository-field");
      ws.set("depends-on-one", workspace);
      ws.set("root", rootWorkspace);
      let errors = check.validate(workspace, ws, rootWorkspace, {
        defaultBranch
      });
      expect(errors.map(({ workspace, ...x }: any) => x))
        .toMatchInlineSnapshot(`
                      Array [
                        Object {
                          "correctRepositoryField": "https://github.com/Thinkmill/manypkg/tree/${defaultBranch}/packages/no-repository-field",
                          "currentRepositoryField": undefined,
                          "type": "INCORRECT_REPOSITORY_FIELD",
                        },
                      ]
                `);

      check.fix(errors[0], {});

      expect((workspace.packageJson as any).repository).toBe(
        `https://github.com/Thinkmill/manypkg/tree/${defaultBranch}/packages/no-repository-field`
      );
    });
    it("should fix root in a different format", () => {
      let ws = getWS();
      let rootWorkspace = getFakeWS("root");
      let defaultBranch = `b${getFakeString(5)}`;

      (rootWorkspace.packageJson as any).repository =
        "https://github.com/Thinkmill/manypkg.git";

      rootWorkspace.dir = __dirname;
      let workspace = getFakeWS("no-repository-field");
      workspace.dir = path.join(__dirname, "packages/no-repository-field");
      ws.set("depends-on-one", workspace);
      ws.set("root", rootWorkspace);
      let errors = check.validate(rootWorkspace, ws, rootWorkspace, {
        defaultBranch
      });
      expect(errors.map(({ workspace, ...x }: any) => x))
        .toMatchInlineSnapshot(`
                      Array [
                        Object {
                          "correctRepositoryField": "https://github.com/Thinkmill/manypkg",
                          "currentRepositoryField": "https://github.com/Thinkmill/manypkg.git",
                          "type": "INCORRECT_REPOSITORY_FIELD",
                        },
                      ]
                `);

      check.fix(errors[0], {});

      expect((rootWorkspace.packageJson as any).repository).toBe(
        "https://github.com/Thinkmill/manypkg"
      );
    });
    it("should do nothing if already in good format", () => {
      let ws = getWS();
      let rootWorkspace = getFakeWS("root");
      let defaultBranch = `b${getFakeString(5)}`;

      (rootWorkspace.packageJson as any).repository =
        "https://github.com/Thinkmill/manypkg";

      rootWorkspace.dir = __dirname;
      let workspace = getFakeWS("no-repository-field");
      workspace.dir = path.join(__dirname, "packages/no-repository-field");
      ws.set("depends-on-one", workspace);
      ws.set("root", rootWorkspace);
      let errors = check.validate(rootWorkspace, ws, rootWorkspace, {
        defaultBranch
      });
      expect(errors.map(({ workspace, ...x }: any) => x)).toMatchInlineSnapshot(
        `Array []`
      );

      expect((rootWorkspace.packageJson as any).repository).toBe(
        "https://github.com/Thinkmill/manypkg"
      );
    });
  });

  describe("azure devops", () => {
    it("should work", () => {
      let ws = getWS();
      let rootWorkspace = getFakeWS("root");
      let defaultBranch = `b${getFakeString(5)}`;

      (rootWorkspace.packageJson as any).repository =
        "https://dev.azure.com/Thinkmill/monorepos/_git/manypkg";
      rootWorkspace.dir = __dirname;
      let workspace = getFakeWS("no-repository-field");
      workspace.dir = path.join(__dirname, "packages/no-repository-field");
      ws.set("depends-on-one", workspace);
      ws.set("root", rootWorkspace);
      let errors = check.validate(workspace, ws, rootWorkspace, {
        defaultBranch
      });
      expect(errors.map(({ workspace, ...x }: any) => x))
        .toMatchInlineSnapshot(`
                      Array [
                        Object {
                          "correctRepositoryField": "https://dev.azure.com/Thinkmill/monorepos/_git/manypkg?path=packages/no-repository-field&version=GB${defaultBranch}&_a=contents",
                          "currentRepositoryField": undefined,
                          "type": "INCORRECT_REPOSITORY_FIELD",
                        },
                      ]
                `);

      check.fix(errors[0], {});

      expect((workspace.packageJson as any).repository).toBe(
        `https://dev.azure.com/Thinkmill/monorepos/_git/manypkg?path=packages/no-repository-field&version=GB${defaultBranch}&_a=contents`
      );
    });
    it("should fix root in a different format", () => {
      let ws = getWS();
      let rootWorkspace = getFakeWS("root");
      let defaultBranch = `b${getFakeString(5)}`;

      (rootWorkspace.packageJson as any).repository =
        "https://Thinkmill@dev.azure.com/Thinkmill/monorepos/_git/manypkg";

      rootWorkspace.dir = __dirname;
      let workspace = getFakeWS("no-repository-field");
      workspace.dir = path.join(__dirname, "packages/no-repository-field");
      ws.set("depends-on-one", workspace);
      ws.set("root", rootWorkspace);
      let errors = check.validate(rootWorkspace, ws, rootWorkspace, {
        defaultBranch
      });
      expect(errors.map(({ workspace, ...x }: any) => x))
        .toMatchInlineSnapshot(`
                      Array [
                        Object {
                          "correctRepositoryField": "https://dev.azure.com/Thinkmill/monorepos/_git/manypkg",
                          "currentRepositoryField": "https://Thinkmill@dev.azure.com/Thinkmill/monorepos/_git/manypkg",
                          "type": "INCORRECT_REPOSITORY_FIELD",
                        },
                      ]
                `);

      check.fix(errors[0], {});

      expect((rootWorkspace.packageJson as any).repository).toBe(
        "https://dev.azure.com/Thinkmill/monorepos/_git/manypkg"
      );
    });
    it("should do nothing if already in good format", () => {
      let ws = getWS();
      let rootWorkspace = getFakeWS("root");
      let defaultBranch = `b${getFakeString(5)}`;

      (rootWorkspace.packageJson as any).repository =
        "https://dev.azure.com/Thinkmill/monorepos/_git/manypkg";

      rootWorkspace.dir = __dirname;
      let workspace = getFakeWS("no-repository-field");
      workspace.dir = path.join(__dirname, "packages/no-repository-field");
      ws.set("depends-on-one", workspace);
      ws.set("root", rootWorkspace);
      let errors = check.validate(rootWorkspace, ws, rootWorkspace, {
        defaultBranch
      });
      expect(errors.map(({ workspace, ...x }: any) => x)).toMatchInlineSnapshot(
        `Array []`
      );

      expect((rootWorkspace.packageJson as any).repository).toBe(
        "https://dev.azure.com/Thinkmill/monorepos/_git/manypkg"
      );
    });
  });
});
