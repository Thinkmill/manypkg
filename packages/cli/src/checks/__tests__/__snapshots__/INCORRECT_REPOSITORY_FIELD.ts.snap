// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`incorrect repository field should work 2`] = `
Array [
  Object {
    "correctRepositoryField": "https://github.com/Thinkmill/manypkg/tree/master/packages/no-repository-field",
    "currentRepositoryField": undefined,
    "type": "INCORRECT_REPOSITORY_FIELD",
    "workspace": Object {
      "config": Object {
        "name": "no-repository-field",
        "version": "1.0.0",
      },
      "dir": "/Users/mitchell/projects/manypkg/packages/cli/src/checks/__tests__/packages/no-repository-field",
      "name": "no-repository-field",
    },
  },
]
`;
